lib/spack/docs/packaging_guide.rst

```diff
@@ -1999,45 +1999,138 @@ the Python extensions provided by them: once for ``+python`` and once
+    |                                    | :py:class:`.Python` extensions   |
+    +------------------------------------+----------------------------------+
+
+
+
+.. note::
+    Choice of the appropriate base class for a package
+        In most cases packagers don't have to worry about the selection of the right base class
+        for a package, as ``spack create`` will make the appropriate choice on their behalf. In those
+        rare cases where manual intervention is needed we need to stress that a
+        package base class depends on the *build system* being used, not the language of the package.
+        For example, a Python extension installed with CMake would ``extends('python')`` and
+        subclass from :py:class:`.CMakePackage`.
+
+^^^^^^^^^^^^^^^^^^^^^
+Installation pipeline
+^^^^^^^^^^^^^^^^^^^^^
+
+When a user runs ``spack install``, Spack:
+
+* Fetches an archive for the correct version of the software
+* Expands the archive
+* Sets the current working directory to the root directory of the expanded archive
+
+Then, depending on the base class of the package under consideration, it will execute
+a certain number of **phases** that reflect the way a package of that type is usually built.
+The name and order in which the phases will be executed can be obtained either reading the API docs in this manual, or
+using the ``spack info`` command:
+
+.. code-block:: console
+    :emphasize-lines: 13,14
 
-``install`` takes a ``spec``: a description of how the package should
-be built, and a ``prefix``: the path to the directory where the
-software should be installed.
+    $ spack info m4
+    AutotoolsPackage:    m4
+    Homepage:            https://www.gnu.org/software/m4/m4.html
 
-Spack provides wrapper functions for ``configure`` and ``make`` so
-that you can call them in a similar way to how you'd call a shell
-command.  In reality, these are Python functions.  Spack provides
-these functions to make writing packages more natural. See the section
-on :ref:`shell wrappers <shell-wrappers>`.
+    Safe versions:
+        1.4.17    ftp://ftp.gnu.org/gnu/m4/m4-1.4.17.tar.gz
+
+    Variants:
+        Name       Default   Description
+
+        sigsegv    on        Build the libsigsegv dependency
+
+    Installation Phases:
+        autoreconf    configure    build    install
+
+    Build Dependencies:
+        libsigsegv
+
+    Link Dependencies:
+        libsigsegv
+
+    Run Dependencies:
+        None
+
+    Virtual Packages:
+        None
+
+    Description:
+        GNU M4 is an implementation of the traditional Unix macro processor.
+
+Typically, phases have default implementations that fit most of the common cases:
+
+.. literalinclude:: ../../../lib/spack/spack/build_systems/autotools.py
+    :pyobject: AutotoolsPackage.configure
+    :linenos:
+
+making it just sufficient for a packager to override a few
+build system specific helper methods or attributes to provide, for instance,
+configure arguments:
```

Comment:
As a reader, I'm learning about install phases, that I can override them, and that I can figure out what they are with `spack info`.  Now I'm told, almost as a side node, that I don't really override install phasees at all, but actually helper methods.  In fact, this is the first time helper methods are introduced.  Uh oh...

Helper methods need to be introduced a lot earlier, with more explanation.  The following questions should be answered:

1. What are helper methods?
2. How do I know what helper methods are available for a base class / phase?  Is this in the API docs too?  Is it in `spack info`?
3. When do I override a helper method vs. the whole phase?